Replace Insight with Bitcoin Core in BlockchainWriter

lautarodragan opened the pull request:
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

Fixes https://github.com/poetapp/node/issues/24
Fixes https://github.com/poetapp/node/issues/93
Partly addresses https://github.com/poetapp/node/issues/293
needed-by https://github.com/poetapp/node/issues/22
needed-by https://github.com/poetapp/node/issues/25
needed-by https://github.com/poetapp/node/issues/27

### Release
 Semantic release is enabled for this repository. Make sure you follow the right commit message convention. 
 We're using semantic-release's default — [Angular Commit Message Conventions](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).

## Description of Changes

This PR replaces the Insight API dependency of the BlockchainWriter module with the Bitcoin Core Wallet, finishing the migration.

### Docker Compose

- A service for regtest was added.
- The testnet service was renamed and its ports were fixed

### Dependencies

`bitcore-lib` and `@types/bitcore-lib` were uninstalled.

### Deleted Tests

Placeholder tests that were introduced to bypass an error in coverage were deleted.

### Rename

This PR started replacing `timestamp` with `anchor`. See https://github.com/poetapp/node/issues/293.

### Configuration

- The default configuration now points to regtest.
- `insightUrl`, `bitcoinAddress` and `bitcoinAddressPrivateKey` were removed.
- No new configuration entries were needed. The onces added in https://github.com/poetapp/node/pull/383 sufficed.

### Fee Estimation

The previous could would always send a fee of `0.001`. This new code is using the [`fundrawtransaction`](https://bitcoincore.org/en/doc/0.16.0/rpc/rawtransactions/fundrawtransaction/) RPC, which estimates fees based on recent blocks.

### How to Test

This is how I tested this change:

- `sudo docker-compose up -d bitcoin-regtest`
- `db.dropDatabase()`
- `npm run build`
- `npm start`
- `npm run test:integration`
- `GET http://localhost:18080/works`
- Verify that response contains a few items.
- After a configured amount of time these items should show `timestamp.ipfsFileHash`, `timestamp.ipfsDirectoryHash` and `transactionId`. These three properties will appear one by one. Keep refreshing until the show.
- `bitcoin-cli generate 1`
- Wait the configured amount of time until BlockchainReader picks up the generated block
- `GET http://localhost:18080/works` again
- The works in the response should now contain a full `timestamp`, with `blockHash`, `blockHeight`, etc.

I ran this test both in regtest and testnet. 

One ideal test I haven't done is node syncing: deploy this code and the dependencies to two different machines and confirm they sync correctly.

## Things Being Left Out

I'm leaving out of this PR a few things since it's actually working as it is and it's blocking several other PRs.

- Update documentation
- https://github.com/poetapp/node/issues/157
- Unit tests
- Possible refactors or code improvements

lautarodragan commented on src/BlockchainWriter/ClaimController.ts:
```diff
@@ -48,90 +45,89 @@ export class ClaimController {
 
-    if (!utxo || !utxo.length)
-      throw new Error(`Wallet seems to be empty. Check funds for ${this.configuration.bitcoinAddress}`)
+  private anchorData = async (data: string) => {
```
I experimented with `pipeP` for this but decided not to use it yet.

I've left the experiments in [another branch](https://github.com/poetapp/node/compare/bitcoin-core-writer...blockchain-writer-functional-experiments?expand=1).

kennylavender replied:
Agreed, it probably requires to much refactoring to use it nicely atm.

kennylavender commented on src/BlockchainWriter/Bitcoin.ts:
```diff
@@ -0,0 +1,2 @@
+export const getData = (prefix: string, version: ReadonlyArray<number>) => (message: string) =>
```
Can/Should this be unit tested?

lautarodragan commented:
Will add proper docs in a new PR.

For now:  in regtest, `generate` should give you funds. Remember to generate at least 101 blocks since mined/coinbase bitcoins cannot be used until they have 100 confirmations.

geoffturk commented on docker-compose.yml:
```diff
@@ -41,21 +41,30 @@ services:
         npm start;
       "
-  poet-bitcoind:
+  bitcoin-testnet:
```
I don't see the need to stand up a testnet docker instance anymore now that we are using regtest. If that is correct, can we either comment this out or remove it? Otherwise when I run `docker up --build` I am starting up a testnet instance and syncing with the testnet blockchain, thus using up bandwidth and storage for no real reason.

kennylavender replied:
Good point @geoffturk. I can't think of a situation that I would still want a testnet docker instance locally.

kennylavender replied:
Actually, now that I thought about it more, I think we do want a mainnet/testnet instance actually for people to run the node on their machines. Maybe we just need a simpler way of expressing which instance we run instead of running all of them, but that can be tackled in a future PR.

lautarodragan replied:
That's correct, I'll split it apart in different docker-compose files. We still want to test stuff in testnet if possible, but `docker up` shouldn't cause you to download the whole testnet.

jwicks31 commented on docker-compose.yml:
```diff
@@ -36,27 +36,20 @@ services:
       - poet-mongo
       - poet-rabbit
       - poet-ipfs
+      - bitcoin
```
should this be bitcoin-regtest?

lautarodragan replied:
yes! good catch.

lautarodragan replied:
fixed!

lautarodragan commented:
@kennylavender @geoffturk @jwicks31 merging now not to block other PRs. I think I've addressed your reviews, but please reach out / raise issues if there's changes you think should be done to this PR.
